managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/EntryImpl.java

```diff
@@ -47,12 +48,22 @@ protected EntryImpl newObject(Handle<EntryImpl> handle) {
     ByteBuf data;
 
     public static EntryImpl create(LedgerEntry ledgerEntry) {
+        return create(ledgerEntry,null);
+    }
+
+    public static EntryImpl create(LedgerEntry ledgerEntry, ManagedLedgerInterceptor managedLedgerInterceptor) {
         EntryImpl entry = RECYCLER.get();
         entry.timestamp = System.nanoTime();
         entry.ledgerId = ledgerEntry.getLedgerId();
         entry.entryId = ledgerEntry.getEntryId();
         entry.data = ledgerEntry.getEntryBuffer();
-        entry.data.retain();
+        ByteBuf cacheBuf = entry.data;
+        if(managedLedgerInterceptor != null)
+            cacheBuf = managedLedgerInterceptor.beforeCacheEntryFromLedger(entry.data);
+        if(cacheBuf != null && cacheBuf != entry.data)
+            entry.data = cacheBuf;
+        else
+            entry.data.retain();
```

Comment:
Instead of checking the equality of the buffer instance, the interceptor should have a well defined semantic on the expectation of the reference count of the returned buffer.

Reply:
Done.